File: backend/pictodaemon/src/processor.ts
// GNU AGPL v3.0
// Written by John Nunley

import sharp, { Sharp } from "sharp";
import {
  SerializedPipeline,
  deserializePipeline,
  TemplateTable,
} from "libnode";
import ImageManager from "./imageManager";

export type ProcessingResult = ProcessingMidResult<Buffer>;

type ProcessingMidResult<T> =
  | {
      type: ProcessingResultType.Success;
      image: T;
    }
  | {
      type: ProcessingResultType.ImageNotFound;
      missing: Array<number>;
    }
  | {
      type: ProcessingResultType.Error;
      error: string;
    };

export enum ProcessingResultType {
  Success,
  ImageNotFound,
  Error,
}

export default async function process(
  pipeline: SerializedPipeline<any>,
  images: ImageManager
): Promise<ProcessingResult> {
  // Deserialize the pipeline.
  const deserialized = deserializePipeline<ProcessingMidResult<Sharp>, any>(
    pipeline,
    templateTable()
  );

  // Set the metadata of all nodes to the image manager.
  for (const node of deserialized.getNodes()) {
    node.setMetadata(images);
  }

  // TODO: Add outputs to the pipeline serialization format.
  const outputId = -1;
  const outputNode = deserialized.getNode(outputId);
  if (outputNode === undefined) {
    return {
      type: ProcessingResultType.Error,
      error: "Output node not found",
    };
  }

  // Get the first output of the output node.
  const outputLink = outputNode.getInputs()[0];
  if (outputLink === undefined) {
    return {
      type: ProcessingResultType.Error,
      error: "Output not found",
    };
  }

  // Get the final output.
  const finalOutput = outputLink.get();
  let imageDescription;
  if (finalOutput.type !== ProcessingResultType.Success) {
    return finalOutput;
  } else {
    imageDescription = finalOutput.image;
  }

  // Process the pipeline.
  const result = await imageDescription.webp().toBuffer();

  return {
    type: ProcessingResultType.Success,
    image: result,
  };
}

function templateTable(): TemplateTable<ProcessingMidResult<Sharp>, any> {
  const table = new TemplateTable<ProcessingMidResult<Sharp>, any>();

  // TODO: Add all the templates here.

  return table;
}

File: backend/ontario-web/ontario-web/processor.py
# GNU AGPL v3 License
# Written by John Nunley
# libnodepy-based pipeline processor, using ontario as a backend.

import json
import os.path as path
import tempfile

from .image_manager import ImageManager
from . import nodes

from ontario import ImageContext, ImageBuilder

from typing import Union

PipelineUnit = Union[ImageBuilder, int]


def process(pipeline: str, images: ImageManager, target: str) -> None:
    """
    Processes a pipeline.
    """

    # Deserialize from JSON
    pipeline = nodes.deserializePipeline(json.loads(pipeline))

    # set metadata for all nodes
    for node in pipeline.getNodes():
        node.setMetadata(PipelineMetadata(images))

    outputNode = pipeline.getOutputNode()
    if outputNode is None:
        raise Exception("No output node.")

    # Get the output
    output = outputNode.getOutputs()[0]
    img = output.getValue()
    if not isinstance(img, ImageBuilder):
        raise Exception("Output is not an image.")

    # Save to file
    img.save_to_file(target).process()


class PipelineMetadata:
    """
    Metadata for a pipeline.
    """

    # Image manager
    images: ImageManager

    def __init__(self, images: ImageManager):
        self.images = images


def make_template_table() -> nodes.TemplateTable[PipelineUnit, PipelineMetadata]:
    table = nodes.TemplateTable()

    # TODO

    return table
